packages/core/src/internal/newTask.js

```diff
@@ -0,0 +1,134 @@
+import deferred from '@redux-saga/deferred'
+import * as is from '@redux-saga/is'
+import { TASK, TASK_CANCEL } from '@redux-saga/symbols'
+import { assignWithSymbols, check, createSetContextWarning } from './utils'
+import { addSagaStack, sagaStackToString } from './error-utils'
+import forkQueue from './forkQueue'
+
+export default function newTask(env, mainTask, parentContext, parentEffectId, meta, isRoot, cont) {
+  let running = true
+  let cancelled = false
+  let aborted = false
```

Comment:
seems like it would be better to make those a single enum, i believe only one should ever be true, smth like
```js
const RUNNING = 0
const CANCELLED = 1
const ABORTED = 2
const DONE = 3

let status = RUNNING
```
should make the trick

Reply:
The problem is that "Tasks can be Cancelled but still Running", so it is impossible that we just use a variable 😞 

FYI, line 155, 156:
https://github.com/redux-saga/redux-saga/blob/a619f15cf46ea463528af86c18fb346e9f275985/packages/core/src/internal/proc.js#L153-L166

Reply:
Not sure when this can happen, maybe this comment is outdated/inaccurate?

Reply:
A task is *running and cancelled* when the task is performing the cleanup logic (code in the finally block).

Reply:
We dont support async cleanups though, so the finally block should return synchronously, therefore it should in theory make it safe to transition from running to cancelled either before making calling into cancellation logic or just after that (preferably former). This could require some minor code changes, but sounds like a reasonable assumption to me that it should be doable, right?

Reply:
Even though there's no async cleanups, a cancelled task can still yield put/select /cancelled effects in their finally block, I'm not sure whether it is okay we just can transit the status from running to cancelled, especially when the put effects in cleanup code maybe delayed due to saga scheduling. 

I'm not saying it is impossible, I just mean that I'm not sure about it. Using a single status variable is attractive, but I think I need more thought on it .